### ISSUE TITLE
Inconsistent behaviour when `Dataset.repeats()` and `max_steps` differ

### ISSUE THREAD

**Issue by elgehelge** [tensorflow/tensorflow#19305]
### System information
(This issue is about an inconsistent behaviour between running TF with and without GPU. I will update the system information when it is clear to me which system that is not behaving. 

### Describe the problem
I get inconsistent behaviour when I try to train a model with/without GPU. The problem can be summarized with these few lines of code:
```
    train_spec = tf.estimator.TrainSpec(
        input_fn=lambda: make_dataset_iter(
            n_repeats=2),  # *** THIS HAS PRECEDENCE ON GPU ***
        max_steps=N_SAMPLES * 3)  # *** THIS HAS PRECEDENCE WITHOUT GPU ***
```
With a `make_dataset_iter` function that produces 20 samples (`N_SAMPLES=20`) I will get a model that is trained for 60 global steps (data repeated 3 times) when I run locally without GPU, and 40 global steps (data repeated 2 times) when I run it in a Docker container on a GPU cluster. Not yet sure whether it is a GPU thing, or something else related to the way I run it in the cloud.

Note that this also is a problem when either `max_steps` or `Dataset.repeats()` are set to the default value `None` and the other is set to something specific. This is probably the use-case where most people will experience this inconsistency.

Which of the two are the expected behaviours?

Full code and logs can be found below.

### Source code / logs
Minimal code reproducing the issue:
```
import tensorflow as tf


N_SAMPLES = 20


def lr_model_fn(features, labels, mode):
    """Simple logistic regression model"""
    prediction = tf.layers.dense(features, units=1)
    loss = tf.losses.mean_squared_error(prediction, labels)

    return tf.estimator.EstimatorSpec(
        mode=mode,
        predictions=prediction,
        loss=loss,
        train_op=tf.train.AdamOptimizer().minimize(
            loss=loss,
            global_step=tf.train.get_global_step(),
        )
    )


def make_dataset_iter(n_repeats=1):
    """Turns an instance of a generator into a Dataset"""
    dataset = tf.data.Dataset.from_generator(
        generator=lambda: zip(range(N_SAMPLES), range(N_SAMPLES)),
        output_types=(tf.float32))
    dataset = dataset.repeat(n_repeats)
    dataset = dataset.make_one_shot_iterator()
    dataset = dataset.get_next()
    features, labels = dataset[0], dataset[1]
    return tf.reshape(features, [-1, 1]), tf.reshape(labels, [-1, 1])


def run_experiment():
    # Create the estimator
    estimator = tf.estimator.Estimator(
        model_fn=lr_model_fn,
    )
    # Build train/eval specs
    train_spec = tf.estimator.TrainSpec(
        input_fn=lambda: make_dataset_iter(
            n_repeats=2),  # *** THIS HAS PRECEDENCE ON GPU ***
        max_steps=N_SAMPLES * 3)  # *** THIS HAS PRECEDENCE WITHOUT GPU ***
    eval_spec = tf.estimator.EvalSpec(
        input_fn=make_dataset_iter)
    # Run training
    tf.estimator.train_and_evaluate(
        estimator,
        train_spec,
        eval_spec
    )


if __name__ == "__main__":
    tf.logging.set_verbosity(tf.logging.INFO)
    run_experiment()

```
[Logs from the model running on GPU cluster](https://github.com/tensorflow/tensorflow/files/2006677/with_gpu_log.txt)
[Logs from my local machine](https://github.com/tensorflow/tensorflow/files/2006678/without_gpu_log.txt)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by elgehelge** [comment]
Yes

**Comment by drpngx** [comment]
@mrry could you take a look?

**Comment by mrry** [comment]
This sounds like a `tf.estimator.TrainSpec` issue, so reassigning to @xiejw.

**Comment by xiejw** [comment]
Sorry, I do not understand the problem here.  What's the inconsistent behavior here? The description and code are not very clear to me. For example

- Where does the BATCH_SIZE have been used in the code? maybe I missed that part. 
- I do not understand these two lines 

            n_repeats=2),  # *** THIS HAS PRECEDENCE ON GPU ***
        max_steps=N_SAMPLES * 3)  # *** THIS HAS PRECEDENCE WITHOUT GPU ***

But as I scanned the example code, I noticed you have a stopping condition in your input_fn, which is based on dataset repeat. 

tf.estimator.train_and_evaluate does not work with that in most of cases.   

As mentioned in the docstring (see [1]), "Stop condition: In order to support both distributed and non-distributed configuration reliably, the only supported stop condition for model training is train_spec.max_steps. If train_spec.max_steps is None, the model is trained forever. " 

It is recommended, for training, make your input_fn returns data for ever, you can do that by using repeat(), i.e., no argument to repeat. And purely using the train_spec.max_steps to control the training length.
 
Hope this works for you.

[1] https://www.tensorflow.org/api_docs/python/tf/estimator/train_and_evaluate

**Comment by elgehelge** [comment]
Thanks for getting back @xiejw . This is not about getting it to work for me - I have no problem with that. This is about reporting an issue to help improving TensorFlow and help other people not banging their heads agains the wall.

Please disregard `BATCH_SIZE`. It was removed in my process of simplifying the example, but apparently I forgot to move the the definition of this. I have removed this line from the example.

I guess the root cause of the supposed inconsistency is whether training is stopped when based on the dataset repeat's `OutOfRangeError` or whether it is based on `train_spec.max_steps`. On one machine it stops when `OutOfRangeError` is reached, on another machine it re-initialises the dataset after getting the `OutOfRangeError` and continues to run until `train_spec.max_steps`.

The question is whether this is a feature or a bug?

It is true that the docstring recommends using only `train_spec.max_steps`, however, this is with distributed training in mind, which this issue is not about.

If you still think this is the desired behaviour, feel free to close the issue.

**Comment by xiejw** [comment]
@elgehelge I totally agree with you and share the same pain. 

I would like to share few things here

1. We are improving the behavior for non-distributed case to avoid such confusion. It is not easy due to some special environments. But we are working on that. You should see some change soon in the master head (should be one or two weeks). With that, we hope this issue can be resolved. 

2. I have opened a team discussion to see whether we could detect the early stopping triggered by the input pipeline, i.e, dataset in your case. At least, a warning in the log should help user. We should give actionable, debuggable instructions. Hope you agree with this. 

I will close this issue once one of both is arrived. Hopefully both.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @xiejw, @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by xiejw** [comment]
Close this issue given the following commit.

https://github.com/tensorflow/tensorflow/commit/3edb609926f2521c726737fc1efeae1572dc6581
